### PR TITLE
Refactoring, fixes and improvements of the line rendering

### PR DESCRIPTION
**Pull request by mvaligursky**
- refactored line rendering
  - fixed issue with the system simply wiping all mesh instances on a layer a line is rendered to
  - proper handling of depthTest (each submitted line batch can specify it instead of first batch settings this for all batches)
  - mesh instance used to render lines is not added/removed from the layer each frame causing expensive LayerComposition update, and uses .visible flag instead
  - moved functionality to its own folder / separated classes to individual files / renamed
- added engine example to show / test the functionality

**New Public API:**
- **drawLine**(start, end, color = Color.WHITE, depthTest = true, layer = ..)
    Draws a single line using a single color, positions specified as Vec3
- **drawLines**(positions, colors, depthTest = true, layer = ..)
    Draws multiple lines, positions specified using Vec3
- **drawLineArrays**(positions, colors, depthTest = true, layer ..)
    Draws multiple lines, positions and colors specified as flat arrays of numbers for performance

**New Private API (subject to change):**
- **drawWireSphere**(center, radius, color = Color.WHITE, segments = 20, depthTest = true, layer = .. )
    Draws a wireframe sphere
- **drawWireAlignedBox**(minPoint, maxPoint, color = Color.WHITE, depthTest = true, layer = ..)
    Draws a wireframe AABB

**Deprecated Public API:**
- **Application.renderLine**
- **Application.renderLines**

Fixes https://github.com/playcanvas/engine/issues/2508
Fixes https://github.com/playcanvas/engine/issues/2069
Discussion: https://github.com/playcanvas/engine/discussions/3073

Screenshot 2021-10-05 at 13 36 37

### CODE REVIEW

File: src/scene/immediate/immediate.js
Comment (Maksims) [inline src/scene/immediate/immediate.js:240]: This does allocations on each render, which would be best to be avoided.

Comment (mvaligursky) [reply src/scene/immediate/immediate.js:240]: yep, was planning to share the array with the function above .. all maybe remove it and share array with the spere and build flat float arrays

Comment (mvaligursky) [reply src/scene/immediate/immediate.js]: and done, thanks!

**Comment by mvaligursky** [comment]
Please discuss here on the API changes
https://github.com/playcanvas/engine/discussions/3073

**Comment by LeXXik** [comment]
Perhaps a rotated cone or cylinder could be added to show how `renderWireCube` is used.

**Comment by slimbuck** [comment]
For complex debug shapes (sphere, cylinder, capsule etc) it's quite limiting not being able to specify an arbitrary transform (especially for physics viz).

In the past I've have taken the approach of building meshes for each shape and then instanced them with matrices (instead of building unique vertices per instance). This has ended up being more efficient overall and quite useful in the more complex debug scenarios.

Wondering how hard this approach would be here? Could we easily add it later (API-wise).

**Comment by mvaligursky** [comment]
I had a version of wired box rendering using a matrix, but at the end decided to remove it .. the problem was that it was rendering a unit sized box, and using world space matrix of an entity with a mesh wasn't working well as it was not taking into account the size of the AABB of the mesh. 

We can add something like this if needed in the future, but I thing generating lines as current implementation is faster due to single render call per layer, instead of render call per primitive.

File: src/scene/immediate/immediate-batch.js
Comment (yaustar) [inline src/scene/immediate/immediate-batch.js:83]: Are we okay with this being a reference to an object or should this be a copy operation?

eg `identityGraphNode.worldTransform.copy(Mat4.IDENTITY)`

**Review by yaustar — APPROVED** [review]
I can't see any obvious issues. May want to wait for a more graphics orientated person?

File: examples/src/examples/graphics/lines.tsx
Comment (yaustar) [inline examples/src/examples/graphics/lines.tsx:190]: Where is the `drawlines` function? I can't see it in the PR

File: src/scene/immediate/immediate-batch.js
Comment (mvaligursky) [reply src/scene/immediate/immediate-batch.js:83]: I'm ok with this, it never changes.

File: examples/src/examples/graphics/lines.tsx
Comment (mvaligursky) [reply examples/src/examples/graphics/lines.tsx:190]: application.js line 1739

Comment (yaustar) [reply examples/src/examples/graphics/lines.tsx:190]: Ah, GitHub collapsed the diff 😓

**Review by slimbuck — APPROVED** [review]
Looking great! Please wait for sign off from @willeastcott before merging, since this PR changes API!

**Comment by willeastcott** [comment]
A few comments from me. This feels like some kind of debug rendering utility library. It's quite opinionated, since there are many ways to render a wire sphere, for example. So my gut feeling is that the shape rendering belongs in the `scripts` folder and 'low-level' line rendering belongs in the core `playcanvas.js`. I can anticipate devs saying 'I want an oriented box function' and 'I want to be able to render line strips' ad infinitum. So that's my concern. But obviously, this is a nice step forward in general so I'm not sure I'd want to block it. 😄

**Comment by mvaligursky** [comment]
I can remove the box / sphere rendering, or at least make them private, and adjust the example to not use them if you think that's a better way for now? I understand your concern, but on the other hand I think basic debug functionality like this is very useful to have easily available.

**Comment by willeastcott** [comment]
What's @slimbuck's thoughts on this?

**Comment by slimbuck** [comment]
From a code structure point of view, I agree that splitting the functionality into low-level immediate mode rendering and higher-level shape drawing would be great. Though this is probably a change we can make at any point.

As for where the high-level shape drawing should live, I'm not exactly sure. If we put them in engine scripts/extras it will help with engine bloat, but then editor users will find it much harder to leverage. Also, using the scripts in es6 applications becomes very difficult because our scripts are not modularised.

I guess the ultimate solution is tree-shaking the engine, but that's a whole other discussion.

**Comment by mvaligursky** [comment]
Based on an offline discussion - the agreement was to make drawWireSphere and drawWireAlignedBox private functions, allowing us to figure out a long term strategy before exposing them in some form.

**Review by willeastcott — APPROVED** [review]

**Pull request merged**
